File: src/main/java/com/dansoftware/libraryapp/util/SystemBrowser.java
package com.dansoftware.libraryapp.util;

import java.awt.*;
import java.io.IOException;
import java.net.URISyntaxException;
import java.net.URL;

/**
 * SystemBrowser is a utility for communicating with the default browser of the system.
 *
 * @author Daniel Gyorffy
 */
public final class SystemBrowser {
    public SystemBrowser() {
    }

    public static boolean isSupported() {
        return Desktop.isDesktopSupported() && Desktop.getDesktop().isSupported(Desktop.Action.BROWSE);
    }

    public void browse(String url) {
        try {
            Desktop.getDesktop().browse(new URL(url).toURI());
        } catch (IOException | URISyntaxException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/java/com/dansoftware/libraryapp/util/SystemBrowser.kt
package com.dansoftware.libraryapp.util

import java.awt.Desktop
import java.net.URL

/**
 * SystemBrowser is a utility for communicating with the default browser of the system.
 *
 * @author Daniel Gyorffy
 */
object SystemBrowser {

    private val browserCore: BrowserCore

    init {
        browserCore = when {
            Desktop.isDesktopSupported() -> BaseBrowser()
            OsInfo.isWindows() -> WindowsBrowser()
            OsInfo.isMac() -> MacBrowser()
            OsInfo.isLinux() -> LinuxBrowser()
            else -> NullBrowser()
        }
    }

    fun browse(url: String) {
        ExploitativeExecutor.submit { browserCore.browse(url) }
    }

    private abstract class BrowserCore {
        abstract fun browse(url: String)
    }

    private class BaseBrowser : BrowserCore() {
        override fun browse(url: String) {
            Desktop.getDesktop().browse(URL(url).toURI())
        }
    }

    private class WindowsBrowser : BrowserCore() {
        override fun browse(url: String) {
            Runtime.getRuntime().also {
                it.exec("rundll32 url.dll,FileProtocolHandler $url")
            }
        }
    }

    private class MacBrowser : BrowserCore() {
        override fun browse(url: String) {
            Runtime.getRuntime().also {
                it.exec("open $url")
            }
        }
    }

    private class LinuxBrowser : BrowserCore() {
        override fun browse(url: String) {
            Runtime.getRuntime().also {
                it.exec("xdg-open $url")
            }
        }
    }

    private class NullBrowser : BrowserCore() {
        override fun browse(url: String) {
        }
    }
}